### ISSUE TITLE
[Internal Error] PAS-NOR-IZE-2330 chpl 1.28.0/1.29.0

### ISSUE THREAD

**Issue by kwaters4** [chapel-lang/chapel#21289]
### Summary of Problem

I managed to stumble on this internal compiler error because of human error. When a function and record/class have the same base name and you have ```foo.bar.t``` instead of```foo.bar_t``` it does not generate the appropriate error message signaling that the function `foo.bar.t``` does not exist. 

### Steps to Reproduce

```chapel
module foo {
    record bar_t{
        var a: int;
    }
    proc bar(x: int){
        return x;
    }
}

foo.bar.t;
// Note this is incorrect and should be;
foo.bar_t;
```

**Compile command:**
```
chpl foo.chpl
```

**Execution command:**
```
./foo
```

**Error:**
```
internal error: PAS-NOR-IZE-2330 chpl version 1.28.0
internal error: PAS-NOR-IZE-2331 chpl version 1.29.0
```

### Configuration Information

- Output of `chpl --version`:
```
chpl version 1.28.0
  built with LLVM version 14.0.6

chpl version 1.29.0
  built with LLVM version 14.0.6

```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: system
CHPL_AUX_FILESYS: none
```

**Comment by mppf** [comment]
Here is a patch that adjusts this error message:

```
diff --git a/compiler/passes/normalize.cpp b/compiler/passes/normalize.cpp
index e1746377f4..a62a1312e6 100644
--- a/compiler/passes/normalize.cpp
+++ b/compiler/passes/normalize.cpp
@@ -2328,7 +2328,7 @@ static void applyGetterTransform(CallExpr* call) {
         call->insertAtHead(gMethodToken);
 
       } else {
-        INT_FATAL(call, "unexpected case");
+        USR_FATAL(call, "cannot use '.' is not supported");
       }
 
     } else if (UnresolvedSymExpr* symExpr = toUnresolvedSymExpr(call->get(2))) {
```

However, I think we might want to give a better error, and to do that we will need to know the type of the thing being accessed with `.`. We could focus on improving this error in the dyno resolver.

**Comment by bradcray** [comment]
I like the idea of changing this from an internal error message to a user one for the time being if that's a step in the right direction, but am not familiar with the context here:  Michael, do you already understand what we're tripping over here?  Guessing, is it that `bar` is known to be a paren-ful function, so `bar` is a FCF expression and we know we can't apply `.` to a FCF without calling it first?

**Comment by mppf** [comment]
@bradcray - we can adjust the error message at the location I noted in my patch, but it's happening during normalize when types aren't really known, so we can't realiably make an error that says e.g. "Can't apply `.` to an FCF" (but we can do that in dyno). IMO we should consider a near-term improvement that is to adjust the wording of the error message to say something like `"This way of using '.' is not currently supported"`. Then, we can focus on making an informative error in dyno.

**Comment by bradcray** [comment]
Sorry, I was trying to ask a more specific question than that:  What specifically is the condition (or conditions) that causes normalize to "know" that it can't apply `.` here (rather than needing to wait until resolution to see what the type of the LHS is, say).  The only case I could come up with was the one I mentioned, but are there others?

I was asking because if there is more than one case, I think it'd be reasonable to simply say "Can't apply `.` to this expression" (where I realize that what "this expression" is could be vague in a line with multiple such expressions, but it's probably enough to get users started, and they could always break the expression across multiple lines to figure it out).  But if there aren't any cases other than "This is a paren-ful FCF, so there's no way to apply to `.` to it," then we could potentially be even more specific.  I couldn't think of any other cases, but also haven't taken the time to look at the code to see if I could guess, so wondered if you might already know.

**Comment by mppf** [comment]
It's weirder than that. The input to `applyGetterTransform` is

```
(183914 CallExpr
  (183915 UnresolvedSymExpr '.')
  (352858 SymExpr 'const-val module=[259]:tmodule=[257]')
  (352857 SymExpr 'foo[183897]:_unknown[54]')
  (352856 UnresolvedSymExpr 'bar')
  (183916 SymExpr "t" 'val _cstr_literal_1547[183912]:c_string[19]'))
```

The code in that function has some conditionals about what kind of Symbol `call->get(2)` refers to. Here it is a Module.

The module token might not make sense here; it is added in resolveModuleCall. 

Here is a patch that gets `applyGetterTransform` to deal with the module token pattern.

```
diff --git a/compiler/passes/normalize.cpp b/compiler/passes/normalize.cpp
index e0b9e6d119..466536019f 100644
--- a/compiler/passes/normalize.cpp
+++ b/compiler/passes/normalize.cpp
@@ -2307,8 +2307,17 @@ static void applyGetterTransform(CallExpr* call) {
   if (call->isNamedAstr(astrSdot)) {
     SET_LINENO(call);
 
-    if (SymExpr* symExpr = toSymExpr(call->get(2))) {
+    // if there is a module token argument, pull it out
+    Expr* modToken = nullptr;
+    Expr* modArg = nullptr;
+    if (SymExpr* symExpr = toSymExpr(call->get(1))) {
+      if (symExpr->symbol() == gModuleToken) {
+        modArg = call->get(2)->remove();
+        modToken = call->get(1)->remove();
+      }
+    }
 
+    if (SymExpr* symExpr = toSymExpr(call->get(2))) {
       symExpr->remove();
 
       if (VarSymbol* var = toVarSymbol(symExpr->symbol())) {
@@ -2346,6 +2355,12 @@ static void applyGetterTransform(CallExpr* call) {
         call->partialTag = true;
       }
     }
+
+    // if there was a module token, put it back in
+    if (modToken) {
+      call->insertAtHead(modArg);
+      call->insertAtHead(modToken);
+    }
   }
 }
```

With that patch, compiling this program outputs

```
bb.chpl:12: In function 'main':
bb.chpl:13: error: 'bar' undeclared (first use this function)
```

which is still not a good error message but at least it's not an internal error.

**Comment by bradcray** [comment]
@mppf: I wanted to check and see whether you've gone forward with your patch at all here, or are hoping I (or someone) will?  Thanks.

**Comment by mppf** [comment]
Nope, I haven't made any more progress on it since then. Just forgot about this. I think the next step is to try testing with my patch above, which is definitely a workaround (so probably needs some comments indicating that by the time it makes it to a PR). I can do that if needed but let's coordinate that part off-issue.

**Comment by bradcray** [comment]
I'm putting a branch together with these patches today and will run a paratest in hopes of improving the situation for 1.30.

**Comment by jabraham17** [comment]
On main, this code no longer results in an internal error and instead gets a pretty decent error message

```
bug.chpl:11: error: unresolved call 'proc(x: int): int.t'
bug.chpl:11: note: because no functions named t found in scope
```

The line `unresolved call 'proc(x: int): int.t'` makes it pretty clear the compiler is trying to call the method `t` on a function.

I am going to open a PR to lock in this error message and resolve this issue
